### ISSUE TITLE
firewalld does not work with the default nftables backend

### ISSUE THREAD

**Issue by puneetse** [clearlinux/distribution#1123]
TL;DR please enable `CONFIG_NF_TABLES_BRIDGE=y` in kernel configs <s>and `RFC3964_IPv4=yes` in  a system-provided `firewalld.conf` for firewalld to work out of box. </s>

[firewalld uses nftables as the default backend](https://firewalld.org/2018/07/nftables-backend). This doesn't work in CL currently. The systemd service starts but the firewall is in a failed state and subsequent `firewall-cmd` commands fail.

```
puneetse@clr-~ $ systemctl status firewalld.service
● firewalld.service - firewalld - dynamic firewall daemon
   Loaded: loaded (/usr/lib/systemd/system/firewalld.service; disabled; vendor preset: disabled)
   Active: active (running) since Mon 2019-08-12 03:57:21 PDT; 3s ago
     Docs: man:firewalld(1)
 Main PID: 15072 (python3)
    Tasks: 2 (limit: 4915)
   Memory: 42.4M
   CGroup: /system.slice/firewalld.service
           └─15072 /usr/bin/python /usr/bin/firewalld --nofork --nopid

clr- systemd[1]: Starting firewalld - dynamic firewall daemon...
clr- systemd[1]: Started firewalld - dynamic firewall daemon.
clr- firewalld[15072]: ERROR: Failed to load '/etc/firewalld/firewalld.conf': [Errno 2] No such file or directory: '/etc/firewalld/firewalld.conf'
clr- firewalld[15072]: WARNING: [Errno 2] No such file or directory: '/etc/firewalld/firewalld.conf'
clr- firewalld[15072]: WARNING: Using fallback firewalld configuration settings.
clr- firewalld[15072]: ERROR: '/usr/bin/ebtables-restore --noflush' failed: ebtables-restore v1.8.3 (nf_tables):
                        line 1: CHAIN_ADD failed (No such file or directory): chain PREROUTING
                        line 1: CHAIN_ADD failed (No such file or directory): chain OUTPUT
                        line 1: CHAIN_ADD failed (No such file or directory): chain POSTROUTING
                        line 2: RULE_FLUSH failed (No such file or directory): rule in chain PREROUTING
                        line 2: RULE_FLUSH failed (No such file or directory): rule in chain OUTPUT
                        line 2: RULE_FLUSH failed (No such file or directory): rule in chain POSTROUTING
                        line 4: CHAIN_ZERO failed (No such file or directory): chain PREROUTING
                        line 4: CHAIN_ZERO failed (No such file or directory): chain OUTPUT
                        line 4: CHAIN_ZERO failed (No such file or directory): chain POSTROUTING
                        line 5: CHAIN_ADD failed (No such file or directory): chain INPUT
                        line 5: CHAIN_ADD failed (No such file or directory): chain FORWARD
                        line 5: CHAIN_ADD failed (No such file or directory): chain OUTPUT
                        line 6: RULE_FLUSH failed (No such file or directory): rule in chain INPUT
                        line 6: RULE_FLUSH failed (No such file or directory): rule in chain FORWARD
                        line 8: C
clr- firewalld[15072]: ERROR: COMMAND_FAILED: '/usr/bin/ebtables-restore --noflush' failed: ebtables-restore v1.8.3 (nf_tables):
                        line 1: CHAIN_ADD failed (No such file or directory): chain PREROUTING
                        line 1: CHAIN_ADD failed (No such file or directory): chain OUTPUT
                        line 1: CHAIN_ADD failed (No such file or directory): chain POSTROUTING
                        line 2: RULE_FLUSH failed (No such file or directory): rule in chain PREROUTING
                        line 2: RULE_FLUSH failed (No such file or directory): rule in chain OUTPUT
                        line 2: RULE_FLUSH failed (No such file or directory): rule in chain POSTROUTING
                        line 4: CHAIN_ZERO failed (No such file or directory): chain PREROUTING
                        line 4: CHAIN_ZERO failed (No such file or directory): chain OUTPUT
                        line 4: CHAIN_ZERO failed (No such file or directory): chain POSTROUTING
                        line 5: CHAIN_ADD failed (No such file or directory): chain INPUT
                        line 5: CHAIN_ADD failed (No such file or directory): chain FORWARD
                        line 5: CHAIN_ADD failed (No such file or directory): chain OUTPUT
                        line 6: RULE_FLUSH failed (No such file or directory): rule in chain INPUT
                        line 6: RULE_FLUSH failed (No such file or directory): rule in chain FORWARD
                        line 8: C

puneetse@clr-~ $ sudo firewall-cmd --state
failed

```

Poking further it seems all ebtables commands return an error:

```
puneetse@clr-~ $ sudo /usr/bin/ebtables -L
Bridge table: filter

Bridge chain: INPUT, entries: 0, policy: ACCEPT

Bridge chain: FORWARD, entries: 0, policy: ACCEPT

Bridge chain: OUTPUT, entries: 0, policy: ACCEPT
ebtables v1.8.3 (nf_tables):  CHAIN_ADD failed (No such file or directory): chain INPUT

puneetse@clr-~ $ echo $?
4
```

Changing the backend to use iptables works around the issue, but nftables is the future for firewalld:

> If for some reason you need to revert to the old iptables backend, you can easily do so by setting FirewallBackend in /etc/firewalld/firewalld.conf to iptables, then restart firewalld. However, please realise that future firewalld development will focus on the nftables backend and not iptables.

Using a custom kernel with `CONFIG_NF_TABLES_BRIDGE=y` seems to resolve the issue 👌, but leaves one final hiccup (upstream bug?):
```
puneetse@clr-~ $ sudo /usr/bin/firewalld --nofork --nopid --debug
2019-08-12 13:05:43 DEBUG1: start()
(snipped)
2019-08-12 13:05:44 DEBUG1: Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/firewall/server/decorators.py", line 53, in handle_exceptions
    return func(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/firewall/server/firewalld.py", line 94, in start
    return self.fw.start()
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 515, in start
    self._start()
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 476, in _start
    self.apply_default_rules(use_transaction=transaction)
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 824, in apply_default_rules
    if self.is_ipv_enabled("ipv6") and self._rfc3964_ipv4:
AttributeError: 'Firewall' object has no attribute '_rfc3964_ipv4'

2019-08-12 13:05:44 Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/firewall/server/decorators.py", line 53, in handle_exceptions
    return func(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/firewall/server/firewalld.py", line 94, in start
    return self.fw.start()
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 515, in start
    self._start()
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 476, in _start
    self.apply_default_rules(use_transaction=transaction)
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 824, in apply_default_rules
    if self.is_ipv_enabled("ipv6") and self._rfc3964_ipv4:
AttributeError: 'Firewall' object has no attribute '_rfc3964_ipv4'

puneetse@clr-~ $ sudo firewall-cmd --state
failed
```

<s>Explicitly adding `RFC3964_IPV4=yes` to `/etc/firewalld/firewalld.conf` fixes this, though it is [supposed to be the default](https://github.com/firewalld/firewalld/blob/master/config/firewalld.conf#L72)</s> See comment below.

```
puneetse@clr-~ $ sudo firewall-cmd --state
running
```

Related to:
https://github.com/clearlinux/distribution/issues/1032
https://github.com/clearlinux/tallow/pull/12
https://github.com/clearlinux/clear-linux-documentation/pull/639

**Comment by ahkok** [comment]
@miguelinux for the kernel config part.

**Comment by ahkok** [comment]
`self.set("RFC3964_IPv4", "yes" if config.FALLBACK_RFC3964_IPV4 else "no")`

Need to figure out what that means, but, it's likely the reason it is disabled by default.

**Comment by puneetse** [comment]
The `RFC3964_IPV4` symptom may be a red herring and caused simply because it is the last loaded config value. The real issue might be the exception behavior when an `/etc` file isn't found.

Just doing an empty `sudo touch /etc/firewalld/firewalld.conf` also worked and took in the default FALLBACK value.

**Comment by puneetse** [comment]
In `firewall/core/fw.py` the logic setting local variables like `self._rfc3964_ipv4 ` to the defaults/fallbacks gets skipped when a conf file doesn't exist because of the try-else pattern. 

I swapped `else` with `finally` to always call this logic, which I think should be okay since the default/fallback values already get set in `firewall/core/io/firewalld_conf.py` but I'd need to test further and I'm not sure if there is another reason to be using try-else here.

```
--- /tmp/firewall/core/fw.py    2019-08-13 04:08:20.927434121 -0700
+++ /usr/lib/python3.7/site-packages/firewall/core/fw.py        2019-08-13 04:09:12.529433408 -0700
@@ -233,25 +233,25 @@

     def _start(self, reload=False, complete_reload=False):
         # initialize firewall
         default_zone = config.FALLBACK_ZONE

         # load firewalld config
         log.debug1("Loading firewalld config file '%s'", config.FIREWALLD_CONF)
         try:
             self._firewalld_conf.read()
         except Exception as msg:
             log.warning(msg)
             log.warning("Using fallback firewalld configuration settings.")
-        else:
+        finally:
             if self._firewalld_conf.get("DefaultZone"):
                 default_zone = self._firewalld_conf.get("DefaultZone")

             if self._firewalld_conf.get("CleanupOnExit"):
                 value = self._firewalld_conf.get("CleanupOnExit")
                 if value is not None and value.lower() in [ "no", "false" ]:
                     self.cleanup_on_exit = False
                 log.debug1("CleanupOnExit is set to '%s'",
                            self.cleanup_on_exit)

             if self._firewalld_conf.get("Lockdown"):
                 value = self._firewalld_conf.get("Lockdown")

(snipped)
            if self._firewalld_conf.get("RFC3964_IPv4"):
                value = self._firewalld_conf.get("RFC3964_IPv4")
                if value.lower() in [ "no", "false" ]:
                    self._rfc3964_ipv4 = False
                else:
                    self._rfc3964_ipv4 = True
                log.debug1("RFC3964_IPv4 is set to '%s'",
                           self._rfc3964_ipv4)
```


`firewall/core/io/firewalld_conf.py`
```
    # load self.filename
    def read(self):
        self.clear()
        try:
            f = open(self.filename, "r")
        except Exception as msg:
            log.error("Failed to load '%s': %s", self.filename, msg)
            self.set("DefaultZone", config.FALLBACK_ZONE)
            self.set("MinimalMark", str(config.FALLBACK_MINIMAL_MARK))
            self.set("CleanupOnExit", "yes" if config.FALLBACK_CLEANUP_ON_EXIT else "no")
            self.set("Lockdown", "yes" if config.FALLBACK_LOCKDOWN else "no")
            self.set("IPv6_rpfilter","yes" if config.FALLBACK_IPV6_RPFILTER else "no")
            self.set("IndividualCalls", "yes" if config.FALLBACK_INDIVIDUAL_CALLS else "no")
            self.set("LogDenied", config.FALLBACK_LOG_DENIED)
            self.set("AutomaticHelpers", config.FALLBACK_AUTOMATIC_HELPERS)
            self.set("FirewallBackend", config.FALLBACK_FIREWALL_BACKEND)
            self.set("FlushAllOnReload", "yes" if config.FALLBACK_FLUSH_ALL_ON_RELOAD else "no")
            self.set("RFC3964_IPv4", "yes" if config.FALLBACK_RFC3964_IPV4 else "no")
            raise
```

**Comment by puneetse** [comment]
> I swapped `else` with `finally` to always call this logic

This works as far as I can tell. Another option would be to fix the individual variable loadingcheck, like the others are wrapped in ` if value in not None:`

I can submit this patch to the CL firewalld package, if no objections.

**Comment by puneetse** [comment]
https://github.com/firewalld/firewalld/pull/509

**Comment by miguelinux** [comment]
The kernel kconfig change will be available at  **linux-5.2.9-824**
